### PR TITLE
Development mode: Push to preview branch

### PR DESCRIPTION
**Pull request by Michkov**
- Adding preview mode which creates separate branch for preview
- Create commit configuring argo-cd to point to forked repository
- Allow to create `hack/preview.env` with custom variables and
  secrets which are not going to be commited but used by preview-mode

The branch with change is cloned and new branch includes configuration for argo-cd.

Workflow of checking branch (regardless of owner):
1. checkout branch feature branch - eg. example
2. run `./hack/preview.sh`

What happen:
1. new branch preview-example is created
2. created commit which configures argocd to point to your fork preview-example
3. branch pushed into your fork
4. argo-cd pointed to preview-example

Your original branch with feature is not touched. In case that you change your feature branch then run preview.sh again.

### CODE REVIEW

File: hack/development-mode.sh [hunk 1]
```diff
@@ -2,19 +2,49 @@
...
-if [ -z "${MY_GIT_REPO_URL}" ]; then
-    MY_GIT_REPO_URL=$(git ls-remote --get-url | sed 's|^git@github.com:|https://github.com/|')
+if [ -f $ROOT/hack/development-mode.env ]; then
+    source $ROOT/hack/development-mode.env
 fi
-if [ -z "${MY_GIT_BRANCH}" ]; then
-    MY_GIT_BRANCH=$(git rev-parse --abbrev-ref HEAD)
+
+if [ -z "$MY_GIT_FORK_REMOTE" ]; then
+    echo "Set MY_GIT_FORK_REMOTE environment to name of your fork remote"
```
Comment (mmorhun) [inline hack/development-mode.sh:10]: Shouldn't we exit here or try to use existing remote if it is the only one?

Comment (Michkov) [reply hack/development-mode.sh:10]: Right, thanks, added

**Review by mmorhun — APPROVED** [review]
LGTM

**Comment by sbose78** [comment]
Is this good for merge?

**Comment by jduimovich** [comment]
Please add a `.gitignore` for `hack/development-mode.env`  to ensure you don't accidentally push to your branch.

**Comment by jduimovich** [comment]
I am a little worried about this workflow.
If I start with my "new-feature" branch and enter devmode, you will delete new-feature-dev and force push the contents of new-feature to dev mode with the overlay change. 
If I want to make changes to my new-feature branch I need to switch back to it, change it, push to it, and re-run the development-mode scripts. 
Normally once you switch the cluster to dev-mode, I usually push changes to the "-dev" branch over and over to update the cluster. If I switch to the non-dev mode branch and re-run development-mode I will lose my changes ?

**Comment by Michkov** [comment]
@jduimovich - `.gitignore` for `hack/development-mode.env` is part of the MR.

The idea here is that you are not working with dev branch at all, only with feature branches.

 `If I want to make changes to my new-feature branch I need to switch back to it, change it, push to it, and re-run the development-mode scripts.`
If you want to make changes to your new-feature branch you need to switch into it, change it, commit changes and re-run develpoment-mode.sh script. Verify the changes and push the feature branch.

` Normally once you switch the cluster to dev-mode, I usually push changes to the "-dev" branch over and over to update the cluster. If I switch to the non-dev mode branch and re-run development-mode I will lose my changes ?`
Now you need to do commits in your feature branch. Instead of push, you need to run deveploment-mode.sh .

The advantage of this is that we do not need to care of reverting `dev` mode changes, since they are in different branch and also in one commit.

**Comment by jduimovich** [comment]
Sorry, missed the .gitignore. Thanks.

The issue in having to switch back to the base feature branch is that you will eventually forget to do it and accidentally pile up changes in the `-dev` branch which then will be lost if do switch back and then run `deveploment-mode.sh`  again. I'm likely to do it as having to remember to switch back every change seems harder than remembering once prior to a PR into upstream.

**Comment by Michkov** [comment]
`-dev` branch is hidden, basically you should not switch manually into it. Maybe I could rename it into `-preview`, to highlight that it's for view and not for development.

**Comment by Michkov** [comment]
Changed branch name to `preview-<feature-branch>`.

**Comment by Michkov** [comment]
@skabashnyuk @sparkoo @sbose78 @goldmann - could you check this workflow if it matches with your desires for development mode? See PR description.

File: hack/development-mode.sh [hunk 2]
```diff
@@ -2,19 +2,50 @@
...
+    exit 1
 fi
 
-#set the local cluster to point to the current git repo and branch and update the path to development
-$ROOT/hack/util-update-app-of-apps.sh $MY_GIT_REPO_URL development $MY_GIT_BRANCH
+MY_GIT_REPO_URL=$(git ls-remote --get-url $MY_GIT_FORK_REMOTE | sed 's|^git@github.com:|https://github.com/|')
+MY_GIT_BRANCH=$(git rev-parse --abbrev-ref HEAD)
+
+
+if echo "$MY_GIT_REPO_URL" | grep -q https://github.com/redhat-appstudio/infra-deployments; then
```
Comment (skabashnyuk) [inline hack/development-mode.sh:18]: is this compatible with ssh based url?

**Comment by skabashnyuk** [comment]
When I work on `infra-deployments` I used to change 3 files.  In this pr I see only changes in `repo-overlay.yaml`. Am I doing something wrong?
Знімок екрана 2022-02-07 о 16 19 31

**Comment by Michkov** [comment]
@skabashnyuk Updated README.md, hope that clarify feature branch vs preview.

Changes to fork can be done on multiple places, but that repo-overlay.yaml is there to have it on one place.

**Review by skabashnyuk — APPROVED** [review]

**Comment by simonbaird** [comment]
FYI I'm using this, and it makes sense to me. Nice.

**Pull request merged**